File: zhengban/pptv.py
import base64
import time
from Crypto.Cipher import PKCS1_v1_5 as Cipher_pkcs1_v1_5
from Crypto.PublicKey import RSA
import math
import random
import execjs
import uuid
import requests
import re


class PPTV:
    def __init__(self,url):
        self.url = "https://oneplay.api.pptv.com/ups-service/play"
        self.video_url = url
        self.pk = '\n'.join([
            '-----BEGIN PUBLIC KEY-----',
            'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqe6XLQF2JmXWgfh09t8TTZsOb6bnj+duiWw4G7pd5Uo1/DN7Xij3Tys9E7XBX0gdXKYI9j+6Fr45bM28fzl4AxUxnhzmbExRt1NJarDGMKo49ViRg1VbL+Wh9kRi+rAxBisdRiP2JEAL+Awqu80chZxxdyoI1k3fSLoZsv/PGkwolE71qsEM4BO1J9RWNp0wlNGqgR+bTwLKkoe7oiZaKaMsSBWNIBDkwgGKFJZzXMXMnqGsDmfbdi32j6hW9DdrxjCx/i9Nzahd1TWVnw9O1AHL5PD5kM3HzqkAewBu38sZxw8DSGYqG0fgVAQtiLHhlD/19F4NKxqL8IVCinMBHQIDAQAB',
            '-----END PUBLIC KEY-----'
        ])
        self.publicKey = RSA.importKey(self.pk)
        self.js = execjs.compile(open("./js/pptv.js").read())
        self.headers = {
            "accept": "*/*",
            "accept-encoding": "gzip, deflate, br",
            "accept-language": "zh-CN,zh;q=0.9",
            "cache-control": "no-cache",
            "referer": "https://v.pptv.com/",
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36"
        }

    def getRamNumber(self, e):
        t = ""
        for i in range(e):
            t += str(hex(math.floor(16 * random.random()))[2:])
        return t.upper()

    def encrypt(self, message):
        cipher = Cipher_pkcs1_v1_5.new(self.publicKey)
        cipher_text = base64.b64encode(cipher.encrypt(message))
        return cipher_text

    def get_cipher(self):
        t = self.getRamNumber(48)
        i = self.getRamNumber(16)
        res = self.encrypt(f"{t},{i}".encode('utf8'))
        encryptParams = self.get_3des(self.get_msg(), t, i)
        return {"cipher": res.decode(), "encryptParams": encryptParams}

    def get_3des(self, msg, key, iv):
        return self.js.call("encrypted", msg, key, iv)

    def get_msg(self):
        js_result = self.js.call("get3rdKeyRandom")
        type = 'mhpptv'
        appId = 'pptv.web.h5'
        appPlt = 'web'
        appVer = '1.0.4'
        channel = 'sn.cultural'
        sdkVer = '1.5.0'
        cid = self.get_cid()
        allowFt = '0,1,2,3'
        rf = '0'
        ppi = '302c393939'
        o = 'www.google.com'
        ahl_ver = '1'
        ahl_random = js_result['random_hex']
        ahl_signa = js_result['signature_hex']
        vvId = uuid.uuid4()
        version = '1'
        https = 'true'
        streamFormat = 3
        result = f"type={type}&appId={appId}&appPlt={appPlt}&appVer={appVer}&channel={channel}&sdkVer={sdkVer}&cid={cid}&allowFt={allowFt}&rf={rf}&ppi={ppi}&o={o}&ahl_ver={ahl_ver}&ahl_random={ahl_random}&ahl_signa={ahl_signa}&vvId={vvId}&version={version}&https={https}&streamFormat={streamFormat}"
        return result

    def get_cid(self):
        res = requests.get(self.video_url, headers=self.headers).text
        cid = re.findall('var webcfg = {"id":(.*?),', res)
        return cid[0]

    def get_sign(self):
        return self.js.call("get3rdKeyRandom")

    def start(self):
        result = self.get_cipher()
        params = {
            "cipher": result.get("cipher"),
            "encryptParams": result.get("encryptParams"),
            "format": "jsonp",
            "cb": f"getPlayEncode_{int(time.time())}"

        }

        res = requests.get(self.url, params=params, headers=self.headers)
        print(res.text)


if __name__ == '__main__':
    PPTV().start()

File: zhengban/js/pptv.js
var CryptoJS = require("crypto-js");
// const msg = "type=mhpptv&appId=pptv.web.h5&appPlt=web&appVer=1.0.4&channel=sn.cultural&sdkVer=1.5.0&cid=23764751&allowFt=0,1,2,3&rf=0&ppi=302c393939&o=www.google.com&ahl_ver=1&ahl_random=56512723466a41555e402b327439736b&ahl_signa=9d6e29f58acde65886d26d364cb0c57145e57fcd8f430fdd3abea0795477937a&vvId=a88a33b5-5644-9435-6e95-b5f34b20c3d6&version=1&https=true&streamFormat=3"
// key = CryptoJS.enc.Hex.parse("3B4E7F4B13980C603A2936C5C3C304C85B823D8A581AA18A"); // key
// iv = CryptoJS.enc.Hex.parse("8C92E8813637F416");

function encrypted (msg, key, iv) {
    return CryptoJS.TripleDES.encrypt(msg, CryptoJS.enc.Hex.parse(key), {
        iv: CryptoJS.enc.Hex.parse(iv),
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7
    }).toString();
}

 function getRandomWithLen () {
    for (var t = "", i = 0; i < 16; i++) {
        var n = 93 * Math.random() + 33 >> 0;
        t += String.fromCharCode(n)
    }
    return t
}
function toHexStr (e) {
    for (var t = "", i = 0; i < e.length; i++)
        t += e.charCodeAt(i).toString(16);
    return t
}

//  function getSignatureWithKey(e){
//     var t = toHexStr(e);
//     console.log(t)
//     return stringify(t)
// }
//  function stringify(e){
//     for (var t = e.words, i = e.sigBytes, n = [], a = 0; a < i; a++) {
//         var r = t[a >>> 2] >>> 24 - a % 4 * 8 & 255;
//         n.push((r >>> 4).toString(16)),
//         n.push((15 & r).toString(16))
//     }
//     return n.join("")
// }
//  function parse(e){
//     for (var t = e.length, i = [], n = 0; n < t; n += 2)
//         i[n >>> 3] |= parseInt(e.substr(n, 2), 16) << 24 - n % 8 * 4;
//     return new s.init(i,t / 2)
// }
function get3rdKeyRandom(){
    var t = "mhpptv"
      , i = getRandomWithLen()
      , n = i;
    i += t + "-1V8oo0Or1f047NaiMTxK123LMFuINTNeI";
    var a = toHexStr(i);
    return {
        random_hex: encodeHex(n),
        signature_hex: a
    }
}
function encodeHex(e) {
    for (var t = "", i = e.length, n = 0; n < i; ++n) {
        var a = e.charCodeAt(n);
        t += "0123456789abcdef"[a >> 4],
        t += "0123456789abcdef"[15 & a]
    }
    return t
}

// const msg = "type=mhpptv&appId=pptv.web.h5&appPlt=web&appVer=1.0.4&channel=sn.cultural&sdkVer=1.5.0&cid=23764751&allowFt=0,1,2,3&rf=0&ppi=302c393939&o=www.google.com&ahl_ver=1&ahl_random=2c6944735c676e643244406266524a7c&ahl_signa=2c6944735c676e643244406266524a7c6d68707074762d3156386f6f304f7231663034374e61694d54784b3132334c4d4675494e544e6549&vvId=1ca862ac-d8eb-4ddd-86a3-daa958004eee&version=1&https=true&streamFormat=3"
// key = "3B4E7F4B13980C603A2936C5C3C304C85B823D8A581AA18A"
// iv = "8C92E8813637F416"
// console.log(encrypted(msg, key, iv).toString())
// console.log(get3rdKeyRandom())